pkg/cloudprovider/vsphereparavirtual/instances.go

```diff
@@ -0,0 +1,257 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package vsphereparavirtual
+
+import (
+	"context"
+	"strings"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/rest"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	cloudprovider "k8s.io/cloud-provider"
+	"k8s.io/klog/v2"
+
+	"k8s.io/cloud-provider-vsphere/pkg/cloudprovider/vsphereparavirtual/vmservice"
+	"k8s.io/cloud-provider-vsphere/pkg/util"
+
+	vmopv1alpha1 "github.com/vmware-tanzu/vm-operator-api/api/v1alpha1"
+)
+
+type instances struct {
+	vmClient  client.Client
+	namespace string
+}
+
+const (
+	providerPrefix = ProviderName + "://"
+)
+
+// DiscoverNodeBackoff is set to be the same with https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/cloud/node_controller.go#L83
+var DiscoverNodeBackoff = wait.Backoff{
+	Steps:    20,
+	Duration: 50 * time.Millisecond,
+	Jitter:   1.0,
+}
+
+func checkError(err error) bool {
+	return err != nil
+}
+
+// discoverNodeByProviderID takes a ProviderID and returns a VirtualMachine if one exists, or nil otherwise
+// VirtualMachine not found is not an error
+func (i instances) discoverNodeByProviderID(ctx context.Context, providerID string) (*vmopv1alpha1.VirtualMachine, error) {
+	var discoveredNode *vmopv1alpha1.VirtualMachine = nil
+
+	// Adding Retry here because there is no retry in caller from node controller
+	// https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/cloud/node_controller.go#L368
+	err := util.RetryOnError(
+		DiscoverNodeBackoff,
+		checkError,
+		func() error {
+			uuid := GetUUIDFromProviderID(providerID)
+			vms := vmopv1alpha1.VirtualMachineList{}
+			err := i.vmClient.List(ctx, &vms, &client.ListOptions{
```

Comment:
likewise here -- we should refactor this bit of code to use dynamic client from client-go instead of controller-runtime.

Reply:
will do this in a follow-up PR